File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    var node: TreeNode?
    var left: TreeNode?
    var right: TreeNode?
    
    
    func inorderSucc(_ node: TreeNode?) -> TreeNode?{
        var cur = node
        while(cur?.left != nil){
            cur = cur?.left
        }
        return cur
    }
    func deleteNode(_ root: TreeNode?, _ key: Int) -> TreeNode? {
       if root == nil{
           return nil
       }
        
        if key == root!.val{
            if root?.left == nil && root?.right == nil{
                return nil
            }
            
            if root?.left == nil{
                return root?.right
            }
            
            if root?.right == nil{
                return root?.left
            }
            
            var sucss = inorderSucc(root?.right)
            root!.val = sucss!.val
            root?.right = deleteNode(root?.right, sucss!.val)
            
            
            
        }else if key > root!.val{
            root?.right = deleteNode(root?.right, key)
        }else if root!.val > key{
            root?.left = deleteNode(root?.left, key)
        }
        
        return root
    }
}

File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* inorderSuccesor(TreeNode *node){
        TreeNode *root = node;
        while(root->left != NULL){
            root = root->left;
        }
        // cout << node->val << endl;
        return root;
    }
    
    TreeNode *delNode(TreeNode *root, int key){
        if(root == NULL) return root;
        
        if(root->val > key){
             root->left = delNode(root->left, key);
        }
         if(root->val < key){
          root->right = delNode(root->right, key);
         }
        
      
        
        if(root->val == key){
            if(root->left != NULL & root->right != NULL){
                TreeNode *node = inorderSuccesor(root->right);
                root->val = node->val;
                int val = node->val;
                TreeNode *temp =  delNode(root->right, val);
                root->right = temp;
                return root;
                
            }else if(root->left == NULL){
                TreeNode *temp = root->right;
                return temp;
            }else {
                TreeNode *temp = root->left;
                return temp;
            }
        }else{
              return root;
        }
        
      
        
        
    }
    
    TreeNode* deleteNode(TreeNode* root, int key) {
        return delNode(root, key);
    }
};